sandbox/samples/cpp/alembicassembly/alembicassembly.cpp

```diff
@@ -0,0 +1,1525 @@
+        {
+            return Model;
+        };
+
+        // Delete this instance.
+        void release() override
+        {
+            delete this;
+        }
+
+        // Expand the contents of the assembly.
+        bool do_expand_contents(
+            const asr::Project&     project,
+            const asr::Assembly*    parent,
+            asf::IAbortSwitch*      abort_switch = 0) override
+        {
+            ///////////////////////////////////////////////////////////////////
+            // retrieve assembly parameters
+            ///////////////////////////////////////////////////////////////////
+            const auto params = get_parameters();
+
+            // file path
+            m_file_path = params.get<std::string>("file_path");
+
+            if (m_file_path.empty())
+            {
+                RENDERER_LOG_ERROR("empty 'file_path' property: %s", get_name());
+                return false;
+            }
+
+            // shutters
+            m_shutter_open = params.get_optional<float>("shutter_open_time", 0.0f);
+            m_shutter_close = params.get_optional<float>("shutter_close_time", 0.0f);
+
+            // time offset
+            m_time_offset = params.get_optional<float>("time_offset", 0.0f);
+
+            // verbose mode
+            m_verbose = params.get_optional<bool>("verbose", false);
+
+            if (m_verbose)
+            {
+                RENDERER_LOG_INFO("file_path: %s\n"
+                                  "shutter_open: %f\n"
+                                  "shutter_close: %f\n"
+                                  "time_offset: %f\n"
+                                  "verbose: %s",
+                    m_file_path.c_str(),
+                    m_shutter_open,
+                    m_shutter_close,
+                    m_time_offset,
+                    m_verbose ? "true" : "false");
+            }
+
+            AbcCoreFac::IFactory factory;
+            AbcCoreFac::IFactory::CoreType core_type;
+
+            Abc::IArchive archive = factory.getArchive(m_file_path, core_type);
+
+            if (!archive.valid())
+            {
+                RENDERER_LOG_WARNING("invalid archive: %s", m_file_path.c_str());
+                return false;
+            }
+
+            if (m_verbose)
+                log_archive(archive, core_type);
+
+            m_xform_seq_stack.push_back(asr::TransformSequence());
+
+            // retrieve archive root object
+            const Abc::IObject& root = archive.getTop();
+
+            if (!root.valid())
+            {
+                RENDERER_LOG_WARNING("invalid root: %s", archive.getName().c_str());
+                return false;
+            }
+
+            if (m_verbose)
+                log_obj(root);
+
+            // root children
+            const size_t children_count = root.getNumChildren();
+
+            if (children_count)
+            {
+                for (size_t i = 0; i < root.getNumChildren(); i++)
```

Comment:
The following idiom guarantees that the upper bound is not reevaluated at each iteration:

    for (size_t i = 0, e = root.getNumChildren(); i < e; i++)

Reply:
Done.

Nice trick!